### ISSUE TITLE
MyPy should report error if variable imported in TYPE_CHECKING block?

### ISSUE THREAD

**Issue by sid-kap** [python/mypy#6104]
Currently, MyPy doesn't report an error when a variable is imported in the `if TYPE_CHECKING` block and then used elsewhere.
For example, MyPy accept this code:
```python
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from typing import NamedTuple

class C(NamedTuple):
    x: int
    y: int
```
even though, at runtime, it fails:
```
Traceback (most recent call last):
  File "type_checking_test.py", line 6, in <module>
    class C(NamedTuple):
NameError: name 'NamedTuple' is not defined
```
(I'm using MyPy version 0.650.)

I think we could fix this by adding an `is_type_checking_only` variable to `SymbolTableNode`, which is True if the variable was imported or assigned within a `TYPE_CHECKING` block, and ensuring that this flag is false for any variables used in places other than type annotations in the actual code.

**Comment by ilevkivskyi** [comment]
> Should this be fixed?

This is highly questionable, TBH. Although implementation is not hard, but still will require some work, while value is relatively small. Also this fails immediately at runtime with a (relatively) clear error.

**Comment by gvanrossum** [comment]
I think it’s worth fixing.

**Comment by JukkaL** [comment]
I agree that this would be worth fixing.

The implementation would have to be clever enough to catch `cast(A, x)` if `A` is not defined at runtime but to allow `cast('A', x)`. 

If `A` is also defined in an `else:` block, it should not be treated as undefined at runtime:

```py
if TYPE_CHECKING:
    from m import A
else:
    class A: pass

class B(A): pass  # Should be OK
```

It would also be nice to require quotes to be used in type annotations (if not using `from __future__ import annotations`).

**Comment by ilevkivskyi** [comment]
@JukkaL 
> It would also be nice to require quotes to be used in type annotations:

We already have https://github.com/python/mypy/issues/948 to track this.

**Comment by NiklasRosenstein** [comment]
I would like to raise this point again.

For a command-line application I am optimising global imports to increase its performance; that often means I need to move imports into a `if TYPE_CHECKING` block and repeat the import at the place in the code where the type is referenced by value. 

Since Mypy doesn't catch that the value is undefined at runtime and I make a mistake of not adding the local import, it results in a `NameError` at runtime that _should_ be really easy for Mypy to catch.

With some pointers, I'm happy to look into a contribution for this fix. (I am very unfamiliar with the Mypy codebase).

**Comment by Viicos** [comment]
Related use case:

`module1` has the following:

```python
if TYPE_CHECKING:
    MyType = int
``` 

`module2`:

```python
# No if TYPE_CHECKING block
from module1 import MyType
```
+ As IDEs usually evaluate `TYPE_CHECKING` to `True`, it's hard to catch the resulting `ImportError`

**Comment by bluenote10** [comment]
@JelleZijlstra To quote from the linked duplicate issue https://github.com/python/mypy/issues/16587#issuecomment-1832410680

> I don't think we should warn here. if TYPE_CHECKING means that the type checker should treat this condition as true, and therefore assume that FunctionType is imported. It's a structured way of lying to the type checker. And if you lie to the type checker, you're responsible for the consequences.

I would argue that this is a special case, because it is a required technique for [breaking import cycles](https://mypy.readthedocs.io/en/stable/runtime_troubles.html#import-cycles) as documented. Without this feature, breaking import cycles cannot be done cleanly, or even the contrary: It is likely to cause bugs!

First a general observation: import cycles are in fact more prominent when using strict typing compared to untyped Python, because even techniques like dependency injection (which typically is intended to break runtime cyclic dependencies) now require imports, but just for the type annotations (i.e., untyped Python would not have these imports). Since import cycles are a more common problem now, it is sensible to provide a clean way to support breaking them.

Now without having a check that a symbol got only imported for type annotation usage, we are actually opening the door for a lot of runtime bugs! In other words, without this feature, the recommendation given in the documentation is actually very likely cause bugs, i.e., "type checks fine, but crashes at runtime" -- which is exactly what we want to prevent with type checking. To demonstrate:

```py
from dataclasses import dataclass, field
from typing import TYPE_CHECKING

# Pure "type import" to break runtime import cycle e.g. for dependency injection.
if TYPE_CHECKING:
    from other_module import Foo

# The following usages are valid, because they only use the symbol in type
# annotation contexts. Note that such usages are common for dependency
# injection.

Alias = list[Foo]

def f1(x: Foo):
    ...

@dataclass
class ClassUsage1:
    foo: Foo

# On the other hand, all the following usages are invalid, because they make
# runtime use of the symbol. Unfortunately the type checker happily accepts
# all these usages in the same module, even though they are all bugs leading
# to crashes at runtime. Ideally the type checker should be aware of the fact
# that the symbol does not exist at runtime, and report all these usages as type
# check errors.

foo = Foo()

def f2(x: Foo = Foo()):
    ...

@dataclass
class ClassUsage2:
    foo: Foo = field(default_factory=lambda: Foo())
```

We've been running into such bugs rather frequently now.

**Comment by Spitfire1900** [comment]
Included another common sample case. In substantial modules, particularly when working with teammates it's easy to get caught in these situations.

It's such a design issue that we flag using `if TYPE_CHECKING:` as a code smell internally and avoid using it.

Unfortunately this has to be addressed as part of an ecosystem-wide discussion as Mypy, Pylance, and Ruff all have the same behavior.

```python
import typing

if typing.TYPE_CHECKING:
    import pathlib


def create_folder(path: pathlib.Path) -> None:
    path.mkdir(parents=True, exist_ok=True)
    

my_path = pathlib.Path("example_folder")
create_folder(my_path)
```
```bash
kyle@mainstay:~$ pipx run ruff check /home/kyle/tmp/6104.py
All checks passed!
kyle@mainstay:~$ pipx run mypy /home/kyle/tmp/6104.py
Success: no issues found in 1 source file
```
```python
Traceback (most recent call last):
  File "/home/kyle/tmp/6104.py", line 11, in <module>
    my_path = pathlib.Path("example_folder")
              ^^^^^^^
NameError: name 'pathlib' is not defined. Did you forget to import 'pathlib'?
```

**Comment by RonnyPfannschmidt** [comment]
IMHO type checking only variables should get a union type in non typing usage

So work fine in type annotation but triggers a error when using in code thats not annotation or guarded

**Comment by sporksmith** [comment]
I already noted this on #16587, but I guess worth noting here too for visibility: [flake8-type-checking](https://pypi.org/project/flake8-type-checking/) seems to catch these problems (TC009)

(Though fwiw I am also +1 for mypy catching this, since it's otherwise a pretty annoying footgun)

**Comment by Spitfire1900** [comment]
@sporksmith nice find on [flake8-type-checking](https://pypi.org/project/flake8-type-checking/)
